File: 0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.java
public class Solution {
    public int shortestSubarray(int[] nums, int k) {
        int n = nums.length;
        long ans = Long.MAX_VALUE;
        Deque<long[]> dq = new LinkedList<>();
        long sum = 0;

        for (int i = 0; i < n; i++) {
            sum += nums[i];
            
            if (sum >= k) {
                ans = Math.min(ans, i + 1);
            }
            
            while (!dq.isEmpty() && sum - dq.peekFirst()[1] >= k) {
                ans = Math.min(ans, i - dq.pollFirst()[0]);
            }
            
            while (!dq.isEmpty() && sum <= dq.peekLast()[1]) {
                dq.pollLast();
            }
            
            dq.addLast(new long[]{i, sum});
        }
        
        return ans == Long.MAX_VALUE ? -1 : (int) ans;
    }
}

File: 0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.cpp
class Solution {
public:
    int shortestSubarray(vector<int>& nums, int k) {
        long n = nums.size(), sum = 0;
        int ans = INT_MAX;
        deque<pair<int, long>>dq;
        for(int i=0; i<n; i++) {
            sum += nums[i];
            if(sum >= k) {
                ans = min(ans, i+1);
            }
            while(!dq.empty() && sum - dq.front().second >= k) {
                ans = min(ans, i - dq.front().first);
                dq.pop_front();
            }
            while(!dq.empty() && sum <= dq.back().second) {
                dq.pop_back();
            }
            dq.push_back({i, sum});
        }
        return ans == INT_MAX ? -1 : ans;
    }
};